### PR TITLE
IGNITE-15829 Request attributes for a service in thin clients.

### PR DESCRIPTION
**Pull request by xtern**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientServicesImpl.java [hunk 1]
```diff
@@ -119,8 +143,10 @@ private ServiceInvocationHandler(String name, long timeout, ClientClusterGroupIm
                 if (nodeIds != null && nodeIds.isEmpty())
                     throw new ClientException("Cluster group is empty.");
 
-                return ch.service(ClientOperation.SERVICE_INVOKE,
-                    req -> writeServiceInvokeRequest(req, nodeIds, method, args),
+                boolean withCallCtx = callCtx != null;
+
+                return ch.service(withCallCtx ? ClientOperation.SERVICE_INVOKE_CALLCTX : ClientOperation.SERVICE_INVOKE,
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientServicesImpl.java:148]: Lets' extend `ClientOperation.SERVICE_INVOKE`, it's redundant to introduce a new client operation type.

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientServicesImpl.java [hunk 2]
```diff
@@ -174,6 +206,8 @@ private void writeServiceInvokeRequest(
                         writer.writeObject(args[i]);
                     }
                 }
+
+                writer.writeMap(withCallCtx ? ((ServiceCallContextImpl)callCtx).values() : null);
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientServicesImpl.java:210]: What if the user implements and passes to method their own `ServiceCallContext` implementation? We should check it and throw an error with reasonable message, or introduce some kind of `attributeNames()` method in public interface, to get all possible attributes for this context.

Comment (alex-plekhanov) [reply modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientServicesImpl.java:210]: Also, should not write anything (including `null`) if `ProtocolBitmaskFeature.SERVICE_INVOKE_CALLCTX` is not supported.

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientServicesImpl.java (hunk 1 above)
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientServicesImpl.java]: Done, thanks!

File: modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientServicesImpl.java (hunk 2 above)
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientServicesImpl.java]: > What if the user implements and passes to method their own `ServiceCallContext` implementation? We should check it and throw an error with reasonable message, or introduce some kind of `attributeNames()` method in public interface, to get all possible attributes for this context.

Added explicit user arguments check.

> Also, should not write anything (including `null`) if `ProtocolBitmaskFeature.SERVICE_INVOKE_CALLCTX` is not supported.

Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/ClientMessageParser.java [hunk 3]
```diff
@@ -277,7 +277,6 @@
     private static final short OP_SERVICE_GET_DESCRIPTOR = 7002;
 
     /** Data streamers. */
-    /** */
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/ClientMessageParser.java]: thanks, done

**Review by alex-plekhanov — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/ClientMessageParser.java [hunk 4]
```diff
@@ -491,7 +490,8 @@ public ClientListenerRequest decode(BinaryReaderExImpl reader) {
                 return new ClientExecuteTaskRequest(reader);
 
             case OP_SERVICE_INVOKE:
-                return new ClientServiceInvokeRequest(reader);
+                return new ClientServiceInvokeRequest(reader,
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/ClientMessageParser.java]: Done

**Pull request closed without merging**